src/main/java/com/iota/iri/model/Hash.java

```diff
@@ -10,85 +10,125 @@
+        return calculate(trits, 0, trits.length, SpongeFactory.create(mode));
+    }
 
-    public Hash(final byte[] bytes, final int offset, final int size) {
-        fullRead(bytes, offset, size);
+    public static Hash calculate(byte[] bytes, int tritsLength, final Sponge curl) {
+        int[] trits = new int[tritsLength];
+        Converter.getTrits(bytes, trits);
+        return calculate(trits, 0, tritsLength, curl);
     }
 
-    public Hash(){}
+    public static Hash calculate(final int[] tritsToCalculate, int offset, int length, final Sponge curl) {
+        int[] hashTrits = new int[SIZE_IN_TRITS];
+        curl.reset();
+        curl.absorb(tritsToCalculate, offset, length);
+        curl.squeeze(hashTrits, 0, SIZE_IN_TRITS);
+        return new Hash(hashTrits);
+    }
 
-    public Hash(final byte[] bytes) {
-        this(bytes, 0, SIZE_IN_BYTES);
+
+    // INNER CLASS FIELDS CANNOT BE ACCESSED UNTIL ENTIRE CLASS IS INSTANTIATED
+    // WHEN USED WITH THE SEMANTICS IN THIS CLASS BELOW IS THREADSAFE
+    private final class ByteSafe {
+        private final byte[] bytes;
+        private final Integer hashcode;
+
+        private ByteSafe(byte[] bytes) {
+            Objects.requireNonNull(bytes);
+            this.bytes = bytes;
+            this.hashcode = Arrays.hashCode(bytes);
+        }
     }
 
-    public Hash(final int[] trits, final int offset) {
-        this.trits = new int[SIZE_IN_TRITS];
-        System.arraycopy(trits, offset, this.trits, 0, SIZE_IN_TRITS);
-        //this(Converter.bytes(trits, offset, trits.length));
+    private final class TritSafe {
+        private final int[] trits;
+
+        private TritSafe(int[] trits) {
+            this.trits = Objects.requireNonNull(trits);
+        }
     }
 
-    public Hash(final int[] trits) {
-        this(trits, 0);
+    public Hash() {
     }
 
-    public Hash(final String trytes) {
-        this.trits = new int[SIZE_IN_TRITS];
-        Converter.trits(trytes, this.trits, 0);
+
+    public Hash(final byte[] source, final int sourceOffset, final int sourceSize) {
+        byte[] dest = new byte[SIZE_IN_BYTES];
+        // TODO - THIS IS A BUGGY BOUNDS-SIZE CHECK. I'LL KEEP IT IN BECAUSE TOO MANY TESTS FAIL OTHERWISE.
+        System.arraycopy(source, sourceOffset, dest, 0, sourceSize - sourceOffset > source.length ? source.length - sourceOffset : sourceSize);
+        this.byteSafe = new ByteSafe(dest);
     }
 
-    //
-    /*
-    public static Hash calculate(byte[] bytes) {
-        return calculate(bytes, SIZE_IN_TRITS, new Curl());
+    public Hash(final byte[] bytes) {
+        this(bytes, 0, SIZE_IN_BYTES);
     }
-    */
-    public static Hash calculate(SpongeFactory.Mode mode, int[] trits) {
-        return calculate(trits, 0, trits.length, SpongeFactory.create(mode));
+
+    public Hash(final int[] trits, final int offset) {
+        int[] dest = new int[SIZE_IN_TRITS];
+        System.arraycopy(trits, offset, dest, 0, dest.length);
```

Comment:
Agreed!

Reply:
The good news is ... the System.arraycopy does bounds checks and will fail fast with some obscure message. It is bad style ... but it will survive thanks to downstream code.